rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,326 @@
+- `node`
+Where the recordings are collected is largely unimportant to the encryption
+strategy, but whether or not they are handled async or sync has different
+requirements.
+
+In sync modes the session recording data is written immediately to long term
+storage without intermediate disk writes. This means we can simply instantiate
+an age encryptor at the start of the session and encrypt the recording data as
+it's sent to long term storage.
+
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to encrypt individual parts, we
+will build a special `io.Writer` that contains a single instance of the age
+encryptor that can proxy writes across multiple files. This will require
+maintaining a concurrency-safe mapping of in-progress uploads to encrypted
+writers which incurs a bit of added complexity. However it intentionally avoids
+any sort of intermediate key management which seems a worthwhile tradeoff.
+
+### Protocols
+
+We record sessions for multiple protocols, including ssh, k8s, database
+sessions and more. Because this approach encrypts at the point of writing
+without modifying the recording structure, the strategy for encryption is
+expected to be the same across all protocols.
+
+### Encryption
+
+At a high level, `age` encryption works by generating a per-file symmetric key
+used for data encryption. That key is wrapped using an asymmetric keypair and
+included in the header section of the encrypted file as a stanza. Plugins
+implementing different key algorithms only affect the crypto involved with
+wrapping and unwrapping data encryption keys.
+
+In both proxy and node recording modes, the public `X25519` key used for
+wrapping `age` data keys will come from
+`session_recording_config.status.active_keys`. All unique public keys present
+will be added as recipients.
+
+### Key Types
+
+This design relies on a few different types of keys.
+- File keys generated by `age`. These are data encryption keys used during
+  symmetric encryption of the recording data.
+- `Identity` and `Recipient` keys used by `age` to wrap and unwrap file keys.
+  These are an asymmetric `X25519` keypair.
+- HSM/KMS-owned `RSA` keypairs used to wrap the `Identity` key. These allow the
+  HSM or KMS keystore to "own" decryption without requiring proxy or host nodes
+  to make encryption requests to the auth server.
+
+### Key Generation
+
+To simplify integration with different HSM/KMS systems, the keypair used for 
+data encryption through `age` will be a software generated `X25519` keypair.
+For the rest of this section I will refer to the public key as the `Recipient`
+and the private key as the `Identity` in keeping with `age` terminology.
+
+The auth server generating the keys will then use the configured CA keystore to
+generate an `RSA` keypair used to encrypt, or wrap, the `Identity`. Once
```

Comment:
will you use OAEP encryption? i'd like to see somewhere a mention of the encryption algorithm and any parameters and how we will do decryption with PKCS11, AWS KMS, GCP KMS